tools/preprocess_pragmas.c

```diff
@@ -0,0 +1,146 @@
+            if (is_in_pragma) {
+                *line_end = '\0';
+                char* version_amount_item = strstr(line, version);
+                if (version_amount_item != NULL) {
+                    if (version_amount_item[len_version] != ':') {
+                        fprintf(stderr, "Found version %s in pragma line but no :amount attached\n", version);
+                        fprintf(stderr, "%s\n", line);
+                        return EXIT_FAILURE;
+                    }
+                    char* version_amount_str_start = &version_amount_item[len_version + 1];
+                    char* version_amount_str_end;
+                    long amount = strtol(version_amount_str_start, &version_amount_str_end, 10);
+                    if (version_amount_str_start == version_amount_str_end) {
+                        fprintf(stderr, "Found version %s in pragma line but no amount integer\n", version);
+                        fprintf(stderr, "%s\n", line);
+                        return EXIT_FAILURE;
+                    }
+                    n_fake_structs = (int)amount;
+                }
+            } else {
+                char* p = line;
+                size_t sz = line_end + 1 - line;
+                while (sz != 0) {
+                    size_t nwritten = fwrite(p, 1, sz, stdout);
+                    if (nwritten == 0) {
+                        fprintf(stderr, "Failed to write to stdout\n");
+                        return EXIT_FAILURE;
+                    }
+                    p += nwritten;
+                    sz -= nwritten;
+                }
+            }
+            if (is_in_pragma && line_end[-1] != '\\') {
+                is_in_pragma = false;
+
+                // Always generate at least one struct,
+                // so that fix_bss.py can know where the increment_block_number pragmas are
+                if (n_fake_structs == 0) {
+                    n_fake_structs = 256;
+                }
+
+                // Write fake structs for BSS ordering
+                // pragma_line_number is used for symbol uniqueness,
+                // and also by fix_bss.py to locate the pragma these symbols originate from.
+                for (int i = 0; i < n_fake_structs; i++)
+                    fprintf(stdout, "struct increment_block_number_%05d_%03d;\n", pragma_line_number, i);
+                fprintf(stdout, "#line %d \"%s\"\n", line_num + 1, filename);
+            }
+            line_num++;
+            if (line_end == last_line_return)
+                break;
+            line = line_end + 1;
+        }
+        assert(bufp <= bufend);
+        assert(bufp > last_line_return);
+        char* next_incomplete_line_start = last_line_return + 1;
+        ptrdiff_t next_incomplete_line_sz = bufp - next_incomplete_line_start;
+        assert(next_incomplete_line_sz >= 0);
+        memmove(buf, next_incomplete_line_start, next_incomplete_line_sz);
+        bufp = buf + next_incomplete_line_sz;
+    }
```

Comment:
It seems like anything between the last newline and the end of the file is not processed, which could be pretty confusing for some poor soul (say if the file ends in `}` with no trailing newline)